From 929150c6abe032a0141d24b961885d47023e7a5f Mon Sep 17 00:00:00 2001
From: desmondwu <desmondwu@ami.com>
Date: Fri, 7 Nov 2025 11:35:38 +0800
Subject: [PATCH] hwmon: Fix NULL pointer dereference in smart_register_psu

pmbus_do_probe(), dependent on the smart_register_psu()
And it might be called before the hwmon smart device is registered.
That will cause an "Unable to handle kernel NULL pointer dereference" error.

Signed-off-by: desmondwu <desmondwu@ami.com>
---
 drivers/hwmon/pmbus/Makefile |  3 +--
 drivers/hwmon/pmbus/smart.c  | 12 ++++++++++--
 2 files changed, 11 insertions(+), 4 deletions(-)

diff --git a/drivers/hwmon/pmbus/Makefile b/drivers/hwmon/pmbus/Makefile
index c38e69e83fb5..d96bf3c2f254 100644
--- a/drivers/hwmon/pmbus/Makefile
+++ b/drivers/hwmon/pmbus/Makefile
@@ -2,10 +2,9 @@
 #
 # Makefile for PMBus chip drivers.
 #
-
+obj-$(CONFIG_SMART_MODULE)      += smart.o
 obj-$(CONFIG_PMBUS)		+= pmbus_core.o
 obj-$(CONFIG_SENSORS_PMBUS)	+= pmbus.o
-obj-$(CONFIG_SMART_MODULE)      += smart.o
 obj-$(CONFIG_SENSORS_ACBEL_FSG032) += acbel-fsg032.o
 obj-$(CONFIG_SENSORS_ADM1266)	+= adm1266.o
 obj-$(CONFIG_SENSORS_ADM1275)	+= adm1275.o
diff --git a/drivers/hwmon/pmbus/smart.c b/drivers/hwmon/pmbus/smart.c
index 16e4b0561368..08277969ca14 100644
--- a/drivers/hwmon/pmbus/smart.c
+++ b/drivers/hwmon/pmbus/smart.c
@@ -1442,8 +1442,16 @@ static struct platform_driver smart_driver = {
 	}
 };
 
-module_platform_driver(smart_driver);
-
+static int __init smart_init(void)
+{
+  return platform_driver_register(&smart_driver);
+}
+static void __exit smart_exit(void)
+{
+  platform_driver_unregister(&smart_driver);
+}
+module_init(smart_init);
+module_exit(smart_exit);
 MODULE_LICENSE("GPL");
 MODULE_AUTHOR("Agata Wiatrowska <agata.wiatrowska@intel.com>");
 MODULE_AUTHOR("Wojciech Dembinski <wojciech.dembinski@intel.com>");
-- 
2.41.0

